<a href="https://colab.research.google.com/github/kivvgsr/data-science/blob/main/handlingimbalance_customer_churn.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [1]:
import requests
import pandas as pd

# Define the Google Drive file ID
file_id = '1k9KKCug7k3mT9adqNDToqIjEsu7J6MtM'

# Define the URL to download the file from Google Drive
url = f'https://drive.google.com/uc?id={file_id}'

# Define the output file name
output = 'dataset.csv'

# Download the file from Google Drive using requests
response = requests.get(url)

# Write the contents to the output file
with open(output, 'wb') as f:
    f.write(response.content)

# Load the dataset using pandas
data = pd.read_csv(output)

# Display the first few rows of the dataset
print(data.head())


   customerID  gender  SeniorCitizen Partner Dependents  tenure PhoneService  \
0  7590-VHVEG  Female              0     Yes         No       1           No   
1  5575-GNVDE    Male              0      No         No      34          Yes   
2  3668-QPYBK    Male              0      No         No       2          Yes   
3  7795-CFOCW    Male              0      No         No      45           No   
4  9237-HQITU  Female              0      No         No       2          Yes   

      MultipleLines InternetService OnlineSecurity  ... DeviceProtection  \
0  No phone service             DSL             No  ...               No   
1                No             DSL            Yes  ...              Yes   
2                No             DSL            Yes  ...               No   
3  No phone service             DSL            Yes  ...              Yes   
4                No     Fiber optic             No  ...               No   

  TechSupport StreamingTV StreamingMovies        Contract Pape

In [2]:
import pandas as pd
from matplotlib import pyplot as plt
import numpy as np
%matplotlib inline

In [3]:

import warnings
warnings.filterwarnings('ignore')

In [4]:
df = data
df.sample(5)

customerID  gender  SeniorCitizen Partner Dependents  tenure  \
3393  4683-WYDOU    Male              0     Yes         No      62   
4976  8165-CBKXO    Male              0     Yes        Yes      66   
1965  3948-XHGNA    Male              0      No         No      24   
3426  4510-HIMLV    Male              0      No         No       1   
5425  4998-IKFSE  Female              0      No         No      30   

     PhoneService     MultipleLines InternetService OnlineSecurity  ...  \
3393          Yes               Yes     Fiber optic             No  ...   
4976           No  No phone service             DSL             No  ...   
1965          Yes                No             DSL            Yes  ...   
3426          Yes                No     Fiber optic             No  ...   
5425          Yes               Yes     Fiber optic             No  ...   

     DeviceProtection TechSupport StreamingTV StreamingMovies        Contract  \
3393              Yes         Yes         Yes             Yes        Two year   
4976              Yes          No         Yes             Yes        Two year   
1965               No          No          No              No  Month-to-month   
3426               No          No          No              No  Month-to-month   
5425              Yes          No         Yes             Yes  Month-to-month   

     PaperlessBilling              PaymentMethod MonthlyCharges  TotalCharges  \
3393              Yes  Bank transfer (automatic)         107.60        6912.7   
4976              Yes    Credit card (automatic)          56.75        3708.4   
1965              Yes  Bank transfer (automatic)          49.55        1210.4   
3426              Yes           Electronic check          69.80          69.8   
5425               No           Electronic check         100.45        3096.9   

     Churn  
3393    No  
4976    No  
1965   Yes  
3426   Yes  
5425    No  

[5 rows x 21 columns]

In [5]:
df.Churn.value_counts()

No     5174
Yes    1869
Name: Churn, dtype: int64

In [6]:

517400/df.shape[0]


73.46301292063042

In [7]:
df.drop('customerID',axis='columns',inplace=True)

In [8]:
df.dtypes

gender               object
SeniorCitizen         int64
Partner              object
Dependents           object
tenure                int64
PhoneService         object
MultipleLines        object
InternetService      object
OnlineSecurity       object
OnlineBackup         object
DeviceProtection     object
TechSupport          object
StreamingTV          object
StreamingMovies      object
Contract             object
PaperlessBilling     object
PaymentMethod        object
MonthlyCharges      float64
TotalCharges         object
Churn                object
dtype: object

## TotalCharges should be float but it is an object. Let's check what's going on with this column

In [9]:
df.TotalCharges.values

array(['29.85', '1889.5', '108.15', ..., '346.45', '306.6', '6844.5'],
      dtype=object)


Ahh... it is string. Lets convert it to numbers

In [10]:
pd.to_numeric(df.TotalCharges,errors = 'coerce').isnull()

0       False
1       False
2       False
3       False
4       False
        ...  
7038    False
7039    False
7040    False
7041    False
7042    False
Name: TotalCharges, Length: 7043, dtype: bool

In [11]:
df[pd.to_numeric(df.TotalCharges,errors = 'coerce').isnull()]

gender  SeniorCitizen Partner Dependents  tenure PhoneService  \
488   Female              0     Yes        Yes       0           No   
753     Male              0      No        Yes       0          Yes   
936   Female              0     Yes        Yes       0          Yes   
1082    Male              0     Yes        Yes       0          Yes   
1340  Female              0     Yes        Yes       0           No   
3331    Male              0     Yes        Yes       0          Yes   
3826    Male              0     Yes        Yes       0          Yes   
4380  Female              0     Yes        Yes       0          Yes   
5218    Male              0     Yes        Yes       0          Yes   
6670  Female              0     Yes        Yes       0          Yes   
6754    Male              0      No        Yes       0          Yes   

         MultipleLines InternetService       OnlineSecurity  \
488   No phone service             DSL                  Yes   
753                 No              No  No internet service   
936                 No             DSL                  Yes   
1082               Yes              No  No internet service   
1340  No phone service             DSL                  Yes   
3331                No              No  No internet service   
3826               Yes              No  No internet service   
4380                No              No  No internet service   
5218                No              No  No internet service   
6670               Yes             DSL                   No   
6754               Yes             DSL                  Yes   

             OnlineBackup     DeviceProtection          TechSupport  \
488                    No                  Yes                  Yes   
753   No internet service  No internet service  No internet service   
936                   Yes                  Yes                   No   
1082  No internet service  No internet service  No internet service   
1340                  Yes                  Yes                  Yes   
3331  No internet service  No internet service  No internet service   
3826  No internet service  No internet service  No internet service   
4380  No internet service  No internet service  No internet service   
5218  No internet service  No internet service  No internet service   
6670                  Yes                  Yes                  Yes   
6754                  Yes                   No                  Yes   

              StreamingTV      StreamingMovies  Contract PaperlessBilling  \
488                   Yes                   No  Two year              Yes   
753   No internet service  No internet service  Two year               No   
936                   Yes                  Yes  Two year               No   
1082  No internet service  No internet service  Two year               No   
1340                  Yes                   No  Two year               No   
3331  No internet service  No internet service  Two year               No   
3826  No internet service  No internet service  Two year               No   
4380  No internet service  No internet service  Two year               No   
5218  No internet service  No internet service  One year              Yes   
6670                  Yes                   No  Two year               No   
6754                   No                   No  Two year              Yes   

                  PaymentMethod  MonthlyCharges TotalCharges Churn  
488   Bank transfer (automatic)           52.55                 No  
753                Mailed check           20.25                 No  
936                Mailed check           80.85                 No  
1082               Mailed check           25.75                 No  
1340    Credit card (automatic)           56.05                 No  
3331               Mailed check           19.85                 No  
3826               Mailed check           25.35                 No  
4380               Mailed check           20.00                 No

In [12]:

df.shape

(7043, 20)

In [13]:
df.iloc[488].TotalCharges

' '

In [14]:
df[df.TotalCharges!=' '].shape

(7032, 20)

In [15]:
df1 = df[df.TotalCharges!=' ']
df1.shape

(7032, 20)

In [16]:
df1.dtypes

gender               object
SeniorCitizen         int64
Partner              object
Dependents           object
tenure                int64
PhoneService         object
MultipleLines        object
InternetService      object
OnlineSecurity       object
OnlineBackup         object
DeviceProtection     object
TechSupport          object
StreamingTV          object
StreamingMovies      object
Contract             object
PaperlessBilling     object
PaymentMethod        object
MonthlyCharges      float64
TotalCharges         object
Churn                object
dtype: object

In [17]:
df1.TotalCharges = pd.to_numeric(df1.TotalCharges)

In [18]:

df1.TotalCharges.values

array([  29.85, 1889.5 ,  108.15, ...,  346.45,  306.6 , 6844.5 ])

In [19]:
df1[df1.Churn=='No']


gender  SeniorCitizen Partner Dependents  tenure PhoneService  \
0     Female              0     Yes         No       1           No   
1       Male              0      No         No      34          Yes   
3       Male              0      No         No      45           No   
6       Male              0      No        Yes      22          Yes   
7     Female              0      No         No      10           No   
...      ...            ...     ...        ...     ...          ...   
7037  Female              0      No         No      72          Yes   
7038    Male              0     Yes        Yes      24          Yes   
7039  Female              0     Yes        Yes      72          Yes   
7040  Female              0     Yes        Yes      11           No   
7042    Male              0      No         No      66          Yes   

         MultipleLines InternetService       OnlineSecurity  \
0     No phone service             DSL                   No   
1                   No             DSL                  Yes   
3     No phone service             DSL                  Yes   
6                  Yes     Fiber optic                   No   
7     No phone service             DSL                  Yes   
...                ...             ...                  ...   
7037                No              No  No internet service   
7038               Yes             DSL                  Yes   
7039               Yes     Fiber optic                   No   
7040  No phone service             DSL                  Yes   
7042                No     Fiber optic                  Yes   

             OnlineBackup     DeviceProtection          TechSupport  \
0                     Yes                   No                   No   
1                      No                  Yes                   No   
3                      No                  Yes                  Yes   
6                     Yes                   No                   No   
7                      No                   No                   No   
...                   ...                  ...                  ...   
7037  No internet service  No internet service  No internet service   
7038                   No                  Yes                  Yes   
7039                  Yes                  Yes                   No   
7040                   No                   No                   No   
7042                   No                  Yes                  Yes   

              StreamingTV      StreamingMovies        Contract  \
0                      No                   No  Month-to-month   
1                      No                   No        One year   
3                      No                   No        One year   
6                     Yes                   No  Month-to-month   
7                      No                   No  Month-to-month   
...                   ...                  ...             ...   
7037  No internet service  No internet service        Two year   
7038                  Yes                  Yes        One year   
7039                  Yes                  Yes        One year   
7040                   No                   No  Month-to-month   
7042                  Yes                  Yes        Two year   

     PaperlessBilling              PaymentMethod  MonthlyCharges  \
0                 Yes           Electronic check           29.85   
1                  No               Mailed check           56.95   
3                  No  Bank transfer (automatic)           42.30   
6                 Yes    Credit card (automatic)           89.10   
7                  No               Mailed check           29.75   
...               ...                        ...             ...   
7037              Yes  Bank transfer (automatic)           21.15   
7038              Yes               Mailed check           84.80   
7039              Yes    Credit card (automatic)          103.20   
7040              Yes           Electronic check           29.60   
70

to print the unique values in object columns to see data values

In [20]:
def print_unique_col_values(df):
       for column in df:
            if df[column].dtypes=='object':
                print(f'{column}: {df[column].unique()}')

In [21]:
print_unique_col_values(df1)

gender: ['Female' 'Male']
Partner: ['Yes' 'No']
Dependents: ['No' 'Yes']
PhoneService: ['No' 'Yes']
MultipleLines: ['No phone service' 'No' 'Yes']
InternetService: ['DSL' 'Fiber optic' 'No']
OnlineSecurity: ['No' 'Yes' 'No internet service']
OnlineBackup: ['Yes' 'No' 'No internet service']
DeviceProtection: ['No' 'Yes' 'No internet service']
TechSupport: ['No' 'Yes' 'No internet service']
StreamingTV: ['No' 'Yes' 'No internet service']
StreamingMovies: ['No' 'Yes' 'No internet service']
Contract: ['Month-to-month' 'One year' 'Two year']
PaperlessBilling: ['Yes' 'No']
PaymentMethod: ['Electronic check' 'Mailed check' 'Bank transfer (automatic)'
 'Credit card (automatic)']
Churn: ['No' 'Yes']


In [22]:
df1.replace('No internet service','No',inplace=True)
df1.replace('No phone service','No',inplace=True)

In [23]:
print_unique_col_values(df1)


gender: ['Female' 'Male']
Partner: ['Yes' 'No']
Dependents: ['No' 'Yes']
PhoneService: ['No' 'Yes']
MultipleLines: ['No' 'Yes']
InternetService: ['DSL' 'Fiber optic' 'No']
OnlineSecurity: ['No' 'Yes']
OnlineBackup: ['Yes' 'No']
DeviceProtection: ['No' 'Yes']
TechSupport: ['No' 'Yes']
StreamingTV: ['No' 'Yes']
StreamingMovies: ['No' 'Yes']
Contract: ['Month-to-month' 'One year' 'Two year']
PaperlessBilling: ['Yes' 'No']
PaymentMethod: ['Electronic check' 'Mailed check' 'Bank transfer (automatic)'
 'Credit card (automatic)']
Churn: ['No' 'Yes']


In [24]:

yes_no_columns = ['Partner','Dependents','PhoneService','MultipleLines','OnlineSecurity','OnlineBackup',
                  'DeviceProtection','TechSupport','StreamingTV','StreamingMovies','PaperlessBilling','Churn']
for col in yes_no_columns:
    df1[col].replace({'Yes': 1,'No': 0},inplace=True)

In [25]:
for col in df1:
    print(f'{col}: {df1[col].unique()}')

gender: ['Female' 'Male']
SeniorCitizen: [0 1]
Partner: [1 0]
Dependents: [0 1]
tenure: [ 1 34  2 45  8 22 10 28 62 13 16 58 49 25 69 52 71 21 12 30 47 72 17 27
  5 46 11 70 63 43 15 60 18 66  9  3 31 50 64 56  7 42 35 48 29 65 38 68
 32 55 37 36 41  6  4 33 67 23 57 61 14 20 53 40 59 24 44 19 54 51 26 39]
PhoneService: [0 1]
MultipleLines: [0 1]
InternetService: ['DSL' 'Fiber optic' 'No']
OnlineSecurity: [0 1]
OnlineBackup: [1 0]
DeviceProtection: [0 1]
TechSupport: [0 1]
StreamingTV: [0 1]
StreamingMovies: [0 1]
Contract: ['Month-to-month' 'One year' 'Two year']
PaperlessBilling: [1 0]
PaymentMethod: ['Electronic check' 'Mailed check' 'Bank transfer (automatic)'
 'Credit card (automatic)']
MonthlyCharges: [29.85 56.95 53.85 ... 63.1  44.2  78.7 ]
TotalCharges: [  29.85 1889.5   108.15 ...  346.45  306.6  6844.5 ]
Churn: [0 1]


In [26]:
df1['gender'].replace({'Female':1,'Male':0},inplace=True)

In [27]:

df1.gender.unique()

array([1, 0])

In [28]:
df2 = pd.get_dummies(data=df1, columns=['InternetService','Contract','PaymentMethod'])
df2.columns

Index(['gender', 'SeniorCitizen', 'Partner', 'Dependents', 'tenure',
       'PhoneService', 'MultipleLines', 'OnlineSecurity', 'OnlineBackup',
       'DeviceProtection', 'TechSupport', 'StreamingTV', 'StreamingMovies',
       'PaperlessBilling', 'MonthlyCharges', 'TotalCharges', 'Churn',
       'InternetService_DSL', 'InternetService_Fiber optic',
       'InternetService_No', 'Contract_Month-to-month', 'Contract_One year',
       'Contract_Two year', 'PaymentMethod_Bank transfer (automatic)',
       'PaymentMethod_Credit card (automatic)',
       'PaymentMethod_Electronic check', 'PaymentMethod_Mailed check'],
      dtype='object')

In [29]:
cols_to_scale = ['tenure','MonthlyCharges','TotalCharges']

from sklearn.preprocessing import MinMaxScaler
scaler = MinMaxScaler()
df2[cols_to_scale] = scaler.fit_transform(df2[cols_to_scale])

In [30]:

for col in df2:
    print(f'{col}: {df2[col].unique()}')

gender: [1 0]
SeniorCitizen: [0 1]
Partner: [1 0]
Dependents: [0 1]
tenure: [0.         0.46478873 0.01408451 0.61971831 0.09859155 0.29577465
 0.12676056 0.38028169 0.85915493 0.16901408 0.21126761 0.8028169
 0.67605634 0.33802817 0.95774648 0.71830986 0.98591549 0.28169014
 0.15492958 0.4084507  0.64788732 1.         0.22535211 0.36619718
 0.05633803 0.63380282 0.14084507 0.97183099 0.87323944 0.5915493
 0.1971831  0.83098592 0.23943662 0.91549296 0.11267606 0.02816901
 0.42253521 0.69014085 0.88732394 0.77464789 0.08450704 0.57746479
 0.47887324 0.66197183 0.3943662  0.90140845 0.52112676 0.94366197
 0.43661972 0.76056338 0.50704225 0.49295775 0.56338028 0.07042254
 0.04225352 0.45070423 0.92957746 0.30985915 0.78873239 0.84507042
 0.18309859 0.26760563 0.73239437 0.54929577 0.81690141 0.32394366
 0.6056338  0.25352113 0.74647887 0.70422535 0.35211268 0.53521127]
PhoneService: [0 1]
MultipleLines: [0 1]
OnlineSecurity: [0 1]
OnlineBackup: [1 0]
DeviceProtection: [0 1]
TechSupport: [

In [31]:
X = df2.drop('Churn',axis='columns')
y = testLabels = df2.Churn.astype(np.float32)

from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=15, stratify=y)

In [32]:
X_train[:10]

gender  SeniorCitizen  Partner  Dependents    tenure  PhoneService  \
684        1              0        0           0  0.000000             1   
2446       1              0        0           0  0.239437             1   
1680       0              0        1           1  0.774648             1   
2220       0              0        1           0  1.000000             1   
2842       1              0        0           0  0.042254             0   
5131       1              0        1           0  0.605634             1   
4372       0              0        1           1  0.000000             1   
4687       0              0        1           0  0.295775             1   
2385       1              0        0           1  0.028169             0   
6773       0              1        1           0  0.507042             1   

      MultipleLines  OnlineSecurity  OnlineBackup  DeviceProtection  ...  \
684               0               0             0                 0  ...   
2446              1               0             1                 0  ...   
1680              1               0             0                 0  ...   
2220              0               1             1                 0  ...   
2842              0               1             0                 1  ...   
5131              1               0             0                 0  ...   
4372              0               0             0                 0  ...   
4687              0               0             0                 1  ...   
2385              0               0             1                 0  ...   
6773              1               0             0                 0  ...   

      InternetService_DSL  InternetService_Fiber optic  InternetService_No  \
684                     0                            1                   0   
2446                    0                            1                   0   
1680                    0                            0                   1   
2220                    1                            0                   0   
2842                    1                            0                   0   
5131                    0                            1                   0   
4372                    0                            0                   1   
4687                    0                            1                   0   
2385                    1                            0                   0   
6773                    0                            1                   0   

      Contract_Month-to-month  Contract_One year  Contract_Two year  \
684                         1                  0                  0   
2446                        1                  0                  0   
1680                        0                  1                  0   
2220                        0                  0                  1   
2842                        1                  0                  0   
5131                        1                  0                  0   
4372                        1                  0                  0   
4687                        1                  0                  0   
2385                        1                  0                  0   
6773                        1                  0                  0   

      PaymentMethod_Bank transfer (automatic)  \
684                                         0   
2446                                        0   
1680                                        0   
2220                                        1   
2842                                        0   
5131                                        0   
4372                                        1   
4687                                        0   
2385                                        0   
6773                                        0   

      PaymentMethod_Credit card (automatic)  PaymentMethod_Electronic check  \
684                                       0                             

In [33]:
!pip install tensorflow-addons


     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 611.8/611.8 kB 3.4 MB/s eta 0:00:00


In [34]:
from tensorflow_addons import losses

In [35]:
import tensorflow as tf
from tensorflow import keras
from sklearn.metrics import confusion_matrix , classification_report


In [36]:
def ANN(X_train, y_train, X_test, y_test, loss, weights):
    model = keras.Sequential([
        keras.layers.Dense(26, input_dim=26, activation='relu'),
        keras.layers.Dense(15, activation='relu'),
        keras.layers.Dense(1, activation='sigmoid')
    ])

    model.compile(optimizer='adam', loss=loss, metrics=['accuracy'])

    if weights == -1:
        model.fit(X_train, y_train, epochs=100)
    else:
        model.fit(X_train, y_train, epochs=100, class_weight = weights)

    print(model.evaluate(X_test, y_test))

    y_preds = model.predict(X_test)
    y_preds = np.round(y_preds)

    print("Classification Report: \n", classification_report(y_test, y_preds))

    return y_preds

In [37]:
y_preds = ANN(X_train, y_train, X_test, y_test, 'binary_crossentropy', -1)

Epoch 1/100
176/176 [==============================] - 4s 3ms/step - loss: 0.5314 - accuracy: 0.7424
Epoch 2/100
176/176 [==============================] - 1s 4ms/step - loss: 0.4381 - accuracy: 0.7836
Epoch 3/100
176/176 [==============================] - 1s 4ms/step - loss: 0.4259 - accuracy: 0.7927
Epoch 4/100
176/176 [==============================] - 1s 4ms/step - loss: 0.4209 - accuracy: 0.7986
Epoch 5/100
176/176 [==============================] - 1s 4ms/step - loss: 0.4166 - accuracy: 0.8050
Epoch 6/100
176/176 [==============================] - 1s 4ms/step - loss: 0.4144 - accuracy: 0.8037
Epoch 7/100
176/176 [==============================] - 1s 4ms/step - loss: 0.4128 - accuracy: 0.8036
Epoch 8/100
176/176 [==============================] - 1s 4ms/step - loss: 0.4107 - accuracy: 0.8078
Epoch 9/100
176/176 [==============================] - 1s 4ms/step - loss: 0.4091 - accuracy: 0.8087
Epoch 10/100
176/176 [==============================] - 1s 3ms/step - loss: 0.4091 - accura

In [38]:

y_test.value_counts()

0.0    1033
1.0     374
Name: Churn, dtype: int64

In [39]:
#class count
count_class_0,count_class_1 = df1.Churn.value_counts()

In [40]:
df_class_0 = df2[df2['Churn']==0]
df_class_1 = df2[df2['Churn']==1]

In [41]:
df_class_0.shape

(5163, 27)

In [42]:
df_class_1.shape

(1869, 27)

#here we can see the imbalace where churn of 0 is 5k and churn for 1 is 1k

# **undersampling**

# lets take only 1869 samples from 0 churn and 1869 samples from nthe churn in order to get the equal no of values

In [43]:

df_class_0_under = df_class_0.sample(n=len(df_class_1), replace=True, random_state=42)


In [44]:
df_class_0_under.shape

(1869, 27)

In [45]:
df_test_under=  pd.concat([df_class_0_under,df_class_1],axis=0)
df_test_under.shape

(3738, 27)

### here this will give the count value of the unique values present in the column

In [46]:
print("Random under - sampling: ")
print(df_test_under.Churn.value_counts())

Random under - sampling: 
0    1869
1    1869
Name: Churn, dtype: int64


In [47]:
x= df_test_under.drop('Churn',axis = 'columns')
y = df_test_under['Churn']

In [48]:

from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(x, y, test_size=0.2, random_state=15, stratify=y)

## here the attribute stratify is mainly used in order to balance the samples

In [49]:
y_train.value_counts() # here because of the stratify in the y_train data we are having the equal no of entities

0    1495
1    1495
Name: Churn, dtype: int64

In [50]:
y_preds = ANN(X_train, y_train, X_test, y_test, 'binary_crossentropy', -1)

Epoch 1/100
94/94 [==============================] - 1s 5ms/step - loss: 0.5876 - accuracy: 0.7214
Epoch 2/100
94/94 [==============================] - 0s 5ms/step - loss: 0.5125 - accuracy: 0.7538
Epoch 3/100
94/94 [==============================] - 0s 5ms/step - loss: 0.4978 - accuracy: 0.7585
Epoch 4/100
94/94 [==============================] - 0s 5ms/step - loss: 0.4937 - accuracy: 0.7579
Epoch 5/100
94/94 [==============================] - 0s 5ms/step - loss: 0.4870 - accuracy: 0.7612
Epoch 6/100
94/94 [==============================] - 0s 4ms/step - loss: 0.4838 - accuracy: 0.7619
Epoch 7/100
94/94 [==============================] - 0s 3ms/step - loss: 0.4813 - accuracy: 0.7622
Epoch 8/100
94/94 [==============================] - 0s 3ms/step - loss: 0.4788 - accuracy: 0.7649
Epoch 9/100
94/94 [==============================] - 0s 3ms/step - loss: 0.4758 - accuracy: 0.7692
Epoch 10/100
94/94 [==============================] - 0s 3ms/step - loss: 0.4735 - accuracy: 0.7659
Epoch 11/

## here our  precission and recall values are inproved

#**METHOD - 2 OVERSAMPLING**

In [51]:
count_class_0,count_class_1

(5163, 1869)

#now lets perform the over sample

In [52]:

df_class_1_over = df_class_1.sample(n=len(df_class_0), replace=True, random_state=42)
df_class_1_over.shape

(5163, 27)

In [53]:
df_test_over = pd.concat([df_class_0,df_class_1_over],axis=0)
df_test_over.shape
print("Random over - sampling: ")
print(df_test_over.Churn.value_counts())

Random over - sampling: 
0    5163
1    5163
Name: Churn, dtype: int64


In [54]:
x= df_test_over.drop('Churn',axis = 'columns')
y = df_test_over['Churn']

In [55]:

from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(x, y, test_size=0.2, random_state=15, stratify=y)

In [56]:
y_train.value_counts() # here because of the stratify in the y_train data we are having the equal no of entities

1    4130
0    4130
Name: Churn, dtype: int64

In [ ]:
y_preds = ANN(X_train, y_train, X_test, y_test, 'binary_crossentropy', -1)

#method 3 : SMOTE
it uses the k nerest

In [58]:
x = df2.drop('Churn',axis='columns')
y = df2['Churn']

In [61]:
y.value_counts()

0    5163
1    1869
Name: Churn, dtype: int64

In [60]:
pip install imbalanced-learn


In [63]:
from imblearn.over_sampling import SMOTE
smote = SMOTE(sampling_strategy='minority')
x_sm, y_sm = smote.fit_resample(x, y)


In [ ]:
x_sm.value_counts(),y_sm.value_counts()

In [75]:

from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(x_sm, y_sm, test_size=0.2, random_state=15, stratify=y_sm)

In [76]:

y_train.value_counts()

1    4130
0    4130
Name: Churn, dtype: int64

In [77]:
y_test.value_counts()

1    1033
0    1033
Name: Churn, dtype: int64

In [ ]:
y_preds = ANN(X_train, y_train, X_test, y_test, 'binary_crossentropy', -1)

#metho 4 : use of ensamble with understanding

In [79]:
df2.Churn.value_counts()

0    5163
1    1869
Name: Churn, dtype: int64

In [80]:
x = df2.drop('Churn',axis='columns')
y = df2['Churn']

In [81]:

from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(x, y, test_size=0.2, random_state=15, stratify=y)

In [82]:
y_train.value_counts()

0    4130
1    1495
Name: Churn, dtype: int64

In [83]:
df3  = X_train.copy()
df3['Churn'] = y_train

In [85]:
df3_class0 =  df3[df3.Churn == 0]
df3_class1 =  df3[df3.Churn == 1]

In [88]:

df3_class0.shape,df3_class1.shape

((4130, 27), (1495, 27))

In [91]:
def get_majority(start,end):
  df_train = pd.concat([df3_class0[start:end],df3_class1], axis = 0)
  x_train = df_train.drop('Churn',axis ='columns')
  y_train =  df_train.Churn
  return x_train , y_train


In [95]:
x_train , y_train =  get_majority(0,1495)
y_preds1 = ANN(x_train, y_train, X_test, y_test, 'binary_crossentropy', -1)

Epoch 1/100
94/94 [==============================] - 2s 5ms/step - loss: 0.6372 - accuracy: 0.6441
Epoch 2/100
94/94 [==============================] - 0s 3ms/step - loss: 0.5177 - accuracy: 0.7612
Epoch 3/100
94/94 [==============================] - 0s 4ms/step - loss: 0.4974 - accuracy: 0.7599
Epoch 4/100
94/94 [==============================] - 0s 5ms/step - loss: 0.4899 - accuracy: 0.7682
Epoch 5/100
94/94 [==============================] - 0s 5ms/step - loss: 0.4862 - accuracy: 0.7709
Epoch 6/100
94/94 [==============================] - 0s 5ms/step - loss: 0.4841 - accuracy: 0.7722
Epoch 7/100
94/94 [==============================] - 0s 5ms/step - loss: 0.4809 - accuracy: 0.7696
Epoch 8/100
94/94 [==============================] - 0s 5ms/step - loss: 0.4804 - accuracy: 0.7709
Epoch 9/100
94/94 [==============================] - 0s 4ms/step - loss: 0.4771 - accuracy: 0.7722
Epoch 10/100
94/94 [==============================] - 1s 5ms/step - loss: 0.4743 - accuracy: 0.7759
Epoch 11/

In [96]:
x_train , y_train =  get_majority(1495,2990)
y_preds2 = ANN(x_train, y_train, X_test, y_test, 'binary_crossentropy', -1)

Epoch 1/100
94/94 [==============================] - 3s 5ms/step - loss: 0.6294 - accuracy: 0.6415
Epoch 2/100
94/94 [==============================] - 0s 5ms/step - loss: 0.5201 - accuracy: 0.7518
Epoch 3/100
94/94 [==============================] - 0s 4ms/step - loss: 0.4977 - accuracy: 0.7609
Epoch 4/100
94/94 [==============================] - 0s 4ms/step - loss: 0.4912 - accuracy: 0.7649
Epoch 5/100
94/94 [==============================] - 0s 4ms/step - loss: 0.4870 - accuracy: 0.7659
Epoch 6/100
94/94 [==============================] - 0s 5ms/step - loss: 0.4830 - accuracy: 0.7645
Epoch 7/100
94/94 [==============================] - 0s 4ms/step - loss: 0.4806 - accuracy: 0.7692
Epoch 8/100
94/94 [==============================] - 0s 4ms/step - loss: 0.4786 - accuracy: 0.7679
Epoch 9/100
94/94 [==============================] - 0s 4ms/step - loss: 0.4770 - accuracy: 0.7686
Epoch 10/100
94/94 [==============================] - 0s 4ms/step - loss: 0.4761 - accuracy: 0.7679
Epoch 11/

In [97]:
x_train , y_train =  get_majority(2990,4130)
y_preds3 = ANN(x_train, y_train, X_test, y_test, 'binary_crossentropy', -1)

Epoch 1/100
83/83 [==============================] - 1s 3ms/step - loss: 0.5971 - accuracy: 0.7123
Epoch 2/100
83/83 [==============================] - 0s 3ms/step - loss: 0.5078 - accuracy: 0.7552
Epoch 3/100
83/83 [==============================] - 0s 3ms/step - loss: 0.4915 - accuracy: 0.7605
Epoch 4/100
83/83 [==============================] - 0s 3ms/step - loss: 0.4856 - accuracy: 0.7674
Epoch 5/100
83/83 [==============================] - 0s 3ms/step - loss: 0.4793 - accuracy: 0.7696
Epoch 6/100
83/83 [==============================] - 0s 4ms/step - loss: 0.4746 - accuracy: 0.7753
Epoch 7/100
83/83 [==============================] - 0s 3ms/step - loss: 0.4714 - accuracy: 0.7742
Epoch 8/100
83/83 [==============================] - 0s 3ms/step - loss: 0.4703 - accuracy: 0.7818
Epoch 9/100
83/83 [==============================] - 0s 3ms/step - loss: 0.4666 - accuracy: 0.7799
Epoch 10/100
83/83 [==============================] - 0s 3ms/step - loss: 0.4649 - accuracy: 0.7806
Epoch 11/

# in this method 4 we take the majority of the votes  if ⁉
vote0 =0
vote1 = 1
vote2 = 0
then majority is 0


if ⁉
vote0=0
vote1=1
vote2 =1
thne majority is 1

henc e we conclude that the if sum of the votes is > 1 then the answer is 1 , else 0

In [99]:
y_ped_final = y_preds1.copy()
for i in range(len(y_preds1)):
  total = y_preds1[i]+y_preds2[i]+y_preds3[i]

  if total > 1 :
    y_ped_final[i] = 1
  else:
    y_ped_final[i] =0


In [102]:
print(classification_report(y_test,y_ped_final))

              precision    recall  f1-score   support

           0       0.91      0.67      0.77      1033
           1       0.47      0.81      0.59       374

    accuracy                           0.70      1407
   macro avg       0.69      0.74      0.68      1407
weighted avg       0.79      0.70      0.72      1407

